codegen support for efficient union representations

vtjnash opened the pull request:
This implements the codegen primitives and return-type calling convention to make moving around (generating, storing, inspecting, returning) inferred union types more efficient.

The representation of the tagged-union is as a pair consisting of a memory pointer of unspecified size and a byte describing its type. The memory pointer may also be a `jl_value_t*` box, which will be reflected in it's type tag. For more extensive details, see the commit message.

vtjnash committed:
codegen: represent Union types via tagged unions

this covers the code-generator support for representing types on the
stack via a tagged type representation

The function `remark_julia_type` now can transform between any two types.
Instead of asserting when asked to drop type information, it'll instead
cast the object to the requested representation.
It'll make boxes, allocate stack copies, and compute tagged unions as
needed to perform the request.

This means that we can use `remark_julia_type` whenever we want type
information to forcibly change.
For example, storing to a slot requires dropping the intermediate type
information and generating a representation suitable for copying into
the memory location. Loading from a slot has the opposite requirement:
it needs to be able to load the slot via it's declared type,
then possibly to cast the result to add type information that may be
used later (to emit optimizations of other intrinsic function calls).

The primitive routines that need to be able to handle tagged unions are:
- remark-type
- load-local
- store-local
- isa
- is
- emit_typeof
- emit_sizeof
- boxed
- unbox
- specsig (not implemented here yet)

Everything else should be possible to handle in inference by using these
primitives to implement union-splitting.

Additionally, the following will be supported once the type layout changes:
- getfield / setfield

The representation of the tagged-union is as a pair
of < union*, selector > memory addresses.
The selector is fixed as a byte, and will union-tag the first 126 isbits.
It records the one-based depth-first count into the type-union of the
isbits objects inside. An index of zero indicates that the union* is
actually a tagged heap-allocated jl_value_t*,
and needs to treated as normal for a boxed object rather than as a
tagged union.

The `union*` memory region may be allocated at *any* size.
The only constraint is that it is big enough to contain the data
currently specified by `selector`.
It might not be big enough to contain the union of all types that
could be stored there according to the associated Union type field.
Use appropriate care when copying.

this also includes improvements for codegen of isa with mixed types

vtjnash committed:
implement a union-return-type calling convention

JeffBezanson commented:
Cool! First request: copy that commit message into the devdocs somewhere :)

johnmyleswhite commented:
So excited to see this being worked on. Am I right to think there are still performance optimizations that can be made? For now, I'm using the following snippet to check the performance of union types:

```
function stable(n)
    s = 0.0
    for i in 1:n
        s += sin(1.0)
    end
    return s
end

function unstable(n)
    s = 0
    for i in 1:n
        s += sin(1.0)
    end
    return s
end

@time stable(10_000_000)
@time unstable(10_000_000)
```

Currently this branch produces very similar behavior as Julia 0.5 did: the union type of `s` in `unstable` causes about a 2x latency degradation and allocates a lot of memory.

andyferris commented:
Very cool, Jameson. I'm guessing from the OP and John's comment that the remaining half of work is to make unions of two isbits types isbits themselves? Would these be laid out be like a variant/`Switch`, or overlapped like a C union?

vtjnash commented:
That's a boring example. I can cut down the allocation count, but they don't really cost any time relative to `sin`. The performance variation seen here is entirely due to a processor bug / feature. If you `@profile` both, you'll see that >99% of time in all versions is spent in the `sin` kernel:
```
julia> @time stable(10_000_000) # comparison
  0.153794 seconds (5 allocations: 176 bytes)
8.414709847530957e6

julia> @time unstable(10_000_000) # altered version of PR
  0.318522 seconds (5 allocations: 176 bytes)
8.414709847530957e6

julia> @time unstable(10_000_000) # master last week
  0.365068 seconds (30.00 M allocations: 457.764 MiB, 3.54% gc time)
8.414709847530957e6

julia> @time unstable(10_000_000) # master today
  0.323274 seconds (10.00 M allocations: 152.588 MiB, 2.94% gc time)
8.414709847530957e6
```

A slightly more interesting version is:
```
julia> @noinline sumup(x::Int) = iseven(x) ? x / 2 : x + 3
sumup (generic function with 1 method)

julia> function f(c)
         x::Int = c
         while x != 1
           x = sumup(x)
         end
       end
f (generic function with 1 method)

julia> @time for i = 1:10^7; f(166138751); end # altered version of PR
  3.253064 seconds

julia> @time for i = 1:10^7; f(166138751); end # master
 11.707397 seconds (650.00 M allocations: 9.686 GiB, 2.39% gc time)
```

Although even here it is very difficult to craft a meaningful benchmark since most of the cost is in using a floating point representation, and the cost of allocation is relatively just noise:

```
julia> @noinline sumup(x) = (xi = unsafe_trunc(Int, x); iseven(xi) ? div(x, 2) : x + 3)
sumup (generic function with 1 method)

julia> function f(c)
         x = c
         while x != 1
           x = sumup(x)
         end
       end

julia> @time for i = 1:10^7; f(166138751); end
  0.977569 seconds

julia> @time for i = 1:10^7; f(166138751.0); end
  4.740574 seconds
```

johnmyleswhite commented:
> That's a boring example. I can cut down the allocation count, but they don't really cost any time relative to `sin`. The performance variation seen here is entirely due to a processor bug / feature. If you `@profile` both, you'll see that >99% of time in all versions is spent in the sin kernel:

I'm confused: if 100% of the cost of evaluating both `stable` and `unstable` was the cost of repeatedly evaluating `sin`, wouldn't the two implementations both finish in `0.15` seconds since they both evaluate `sin` 10,000,000 times and only differ in steps other than the evaluation of `sin`? My mental model is that the cost should be `total_time = time_in_sin + time_other` and that the two functions should share `time_in_sin`, so there's clearly a lot of time spent in other steps, even if those steps aren't allocations as your example shows. What am I missing from that model of the cost of evaluating `unstable`?

andyferris commented:
To be honest, I am also a little unclear what those benchmarks are showing...

(also - what is "altered version of PR"? is that work on the isbits case - you say "I can cut down the allocation count"?)

vtjnash commented:
> My mental model is that the cost should be ...

That cost model only works for macro scale. In this case, the cost model is basically `P(processor detects the loop) + P(register stalls)`. And yes, that a probability estimate, not a performance characteristic. There's some factors that are roughly controllable to increase P (fewer branches, smaller loop, predictable memory accesses, the implementation of register dependencies / size of out-of-order pipeline), but it's not really entirely controllable (only Intel knows the implementation details, and they haven't shared anything. although we know that it differs very widely by processor microarchitecture family and node).

(yeah, the altered case is some test code I put together that does some inference re-arrangements to prevent allocation in this case in exchange for worse code in the general case. Since it's just inference work, it doesn't really matter for this PR – I could have gotten the same result just by writing out some of the extra branches in those function manually).

andyferris commented:
Interesting.

Is there some more generic cases that we can test with this branch? Maybe `DataArrays` would be a good example of the kind of things this PR will solve, but it seems to be broken on v0.6.

vchuravy commented:
On 32bit windows: 
```
julia: /home/travis/build/JuliaLang/julia/src/codegen.cpp:503: jl_cgval_t::jl_cgval_t(const jl_cgval_t&, jl_value_t*, llvm::Value*): Assertion `isboxed || v.typ == typ || tindex' failed.
```
I could spend a few cycles looking into this if we are pushing this for v0.6.

andyferris commented:
Hmm... Is this a regression? My benchmark was a bit strange. Let's call this test "DataArrays lite":

```julia
immutable NA; end
Base.:+(::NA, ::Int) = NA()
Base.:+(::NA, ::NA) = NA()
Base.:+(::Int, ::NA) = NA()
n = 10_000_000
v1 = Vector{Union{Int,NA}}(n)
v2 = Vector{Union{Int,NA}}(n)
for i = 1:n
    if rand() < 0.6666
        v1[i] = rand(1:100)
    else
        v1[i] = NA()
    end
    if rand() < 0.6666
        v2[i] = rand(1:100)
    else
        v2[i] = NA()
    end
end
v1 + v2
```

And the benchmark results:
```julia
julia> @time v1 + v2;  # v0.5.0
  0.439152 seconds (10.00 M allocations: 228.874 MB, 24.30% gc time)

julia> @time v1 + v2;  # master
  0.510497 seconds (10.00 M allocations: 305.180 MiB, 23.54% gc time)

julia> @time v1 + v2;  # this PR
  0.747177 seconds (10.00 M allocations: 228.887 MiB, 17.05% gc time)
```
Does this make sense?

When the isbits stuff is done, I was expecting this to be with a factor of two of the `Vector{Int}` speed (0.031480 seconds (7 allocations: 76.294 MB, 6.53% gc time)).

andyferris commented:
And finally, by using/extending nullables on v0.5:
```julia
function Base.:+(i1::Nullable{Int}, i2::Nullable{Int}) 
    if isnull(i1) || isnull(i2)
        return Nullable{Int}()
    else
        return Nullable(get(i1) + get(i2))
    end
end
v1 = Vector{Nullable{Int}}(n)
v2 = Vector{Nullable{Int}}(n)
v3 = Vector{Nullable{Int}}(n)
for i = 1:n
    if rand() < 0.6666
        v1[i] = Nullable(rand(1:100))
    else
        v1[i] = Nullable{Int}()
    end
    if rand() < 0.6666
        v2[i] = Nullable(rand(1:100))
    else
        v2[i] = Nullable{Int}()
    end
end
```
yields
```julia
julia> @time map!(+, v3, v1, v2);
  0.132412 seconds (4 allocations: 160 bytes)
```
which is about 6 times slower than the `Int` case and about 3 times faster than this PR currently (I didn't attempt to optimize that kernel for `+`).

andreasnoack commented:
> Maybe DataArrays would be a good example of the kind of things this PR will solve, but it seems to be broken on v0.6.

You can try this PR https://github.com/JuliaStats/DataArrays.jl/pull/235

vtjnash commented:
@vchuravy I already looked at it – the assertion is wrong. I just need to push a fix to avoid calling it there.

@andyferris This PR is only for the implementation of the codegen support. It doesn't implement support for representing unions efficiently in vectors.

andyferris commented:
>  This PR is only for the implementation of the codegen support. It doesn't implement support for representing unions efficiently in vectors.

Right - I'm getting ahead of you (because this is exciting :smile: )

vtjnash committed:
distinguish between "must" and "may" when remarking julia type information

expr-type and slot-type may improve type information (but try not to destroy it)
while slot-assignment and return must convert the type information (even if it loses information / forces boxing)

vtjnash committed:
change union tindex representation to have the high bit indicate boxed

this allows (tindex & 0x7f) to always be exact,
saving the need for any dynamic loads in many cases

vtjnash committed:
fix over-pessimistic emit_isa for Type

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

vtjnash committed:
ensure the value in the gcroot is transferred during union conversion

convert_julia_type is only ever called with needsgcroot=false,
so this is mostly dead-code right now

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20593#issuecomment-280410062) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/1ec20e0f6550380132d7d041c7220de0e7c87cb3/c9132d2_vs_7ab6428/report.md). cc @jrevels

vtjnash commented:
Wow cool, I've never created a 14552.04% slowdown before. I didn't even have to allocate anything!

vtjnash committed:
ensure that constants are treated as values rather than opaque pointers

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20593#issuecomment-280584323) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d7fec493d2618084e99db6e94762fea88a3f05f7/ed76ab0_vs_3958491/report.md). cc @jrevels

vtjnash commented on src/codegen.cpp:
```diff
@@ -498,8 +525,9 @@ struct jl_cgval_t {
-    jl_cgval_t value; // a value, if the var is unboxed or SSA (and thus boxroot == NULL)
+    Instruction *boxroot; // an address, if the var might be in a jl_value_t** stack slot (marked tbaa_const, if appropriate)
+    jl_cgval_t value; // a stack slot or constant value
+    Value *pTIndex; // where the current value is stored
```
this has the wrong comment

vtjnash committed:
better optimize code for copying a value to an unboxed union

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/20593#issuecomment-281280831) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d3eee15dc708a4bd4fd8569d82652f801da11d5c/a13d81e_vs_e21caee/report.md). cc @jrevels

JeffBezanson commented:
Based on a discussion with @vtjnash yesterday, my understanding was that there was a ~10% slowdown in the raytracer benchmark (in a marginal case involving type-unstable code) due to the new calling convention for functions with Union return types. However it looks like the latest comment has fixed that. If so, that's awesome and this LGTM.

tkelman commented:
what about the 2.4x slowdowns on sqrtm?

JeffBezanson commented:
Yes, good point. We should look into that.

vtjnash commented:
Agreed. But it's unrelated to this PR: it's also shown up in the benchmarks for the past three dailies (https://github.com/JuliaCI/BaseBenchmarkReports/blob/c8c6e85751283542fa8534d17b3b6ea3c637ddcc/daily_2017_2_20/report.md, https://github.com/JuliaCI/BaseBenchmarkReports/blob/400a9f263c95328201eb29616964bb351e574667/daily_2017_2_19/report.md, https://github.com/JuliaCI/BaseBenchmarkReports/blob/2b74ebda5dfabf19cb5b95530482185720702476/daily_2017_2_18/report.md)

The slowdown / speedup was not being reported by `@profile`, so I assumed it was related to the union, it was not. It was a little hard for profile to attribute that the cost was due to an extra gc-root in a function (10% more!). I already squashed that fix in. That last commit also helps a little, but mostly just generates nicer code for the benchmarks.

JeffBezanson commented:
Not related to the changes here, but I'm confused by the comment on the `tbaa` field of `cgval_t`:

```
    MDNode *tbaa; // The related tbaa node. Non-NULL iff this is not a pointer.
    bool ispointer() const
    {
        return tbaa != nullptr;
    }
```

The comment says non-NULL if this is *not* a pointer, but when it's non-null `ispointer` returns true! My guess is that the function should be called something like `hasaddress` instead? `ispointer` makes me think the value is represented as a `jl_value_t*`.

vtjnash commented:
I think there's just an extra `not` in that sentence that was not intended. A `jl_value_t` would have `ispointer` and `isboxed` set. It tells you that you can directly ask for the address of the value / field of the value.

vtjnash committed:
add union-splitting commit comment to devdocs

starting a document describing the overall code-generator structure

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+with names, via the usual symbol resolution mechanism in the linker.
+
+Note too that ccall functions are also handled separately,
+via a manual GOT + PLT.
```
spell out acronyms?

vtjnash replied:
standard jargon exists for a reason

tkelman replied:
what, to confuse people? this is documentation, very welcome to have it at all, just asking for it to be made slightly clearer

vtjnash replied:
because Global Offset Table is not actually clearer, just longer

tkelman replied:
yes it is clearer - I can look that up and go find out more about that if it's clear that's what's being referred to, unlike GOT on its own without any explanation where first the reader needs to figure out what the acronym stands for

tkelman replied:
it's completely standard practice to spell acronyms out the first time you use them, then use the acronym from then on if you must http://blog.apastyle.org/apastyle/abbreviations/#Q1

otherwise the term serves no purpose to anyone who doesn't know what you're referring to

ararslan replied:
There are also folks like me who don't have a solid background with this stuff; I didn't know that GOT was Global Object Table, and I still don't know what PLT is. Big +1 for spelling out acronyms, at least the first time they're used. I think that would help a lot.

StefanKarpinski replied:
@vtjnash: In any writing, you should spell out the first usage, and put the abbreviation in parens. From then on in the same document you may refer to it by abbreviation. Claiming that there's justification for doing anything else is wrong and/or lazy. If you're feeling extra helpful to your reader, you can also link to the wikipedia article on the topic from the first (spelled out) usage of the term.

StefanKarpinski replied:
That said, it's better to have this documentation and others can fix up the formatting and writing as long as we know what it means. I would not have known what GOT stood for without asking you, however. PLT I happen to know.

KristofferC replied:
GOT is a pretty standard acronym though... Game of Thrones of course.

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+`mark_julia_type` (for immediate values) and `mark_julia_slot` (for pointers to values).
+
+The function `convert_julia_type` can transform between any two types.
+When it returns and returns an `cgval.typ` set to `typ`.
```
unclear - reword?

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+primitives to implement union-splitting.
+
+The representation of the tagged-union is as a pair
+of < void* union, byte selector >.
```
code highlight ?

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+It records the one-based depth-first count into the type-union of the
+isbits objects inside. An index of zero indicates that the `union*` is
+actually a tagged heap-allocated `jl_value_t*`,
+and needs to treated as normal for a boxed object rather than as a
```
needs to be treated as normal

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+tagged union.
+
+The high bit of the selector (`byte & 0x80`) can be tested to determine if the
+`void*` if the `void*` is actually a heap-allocated box,
```
duplicate "if the `void*`"

tkelman commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+- Tuples of VecElement types get passed in vector registers.
+- Structs get passed on the stack.
+- Return values are handle similarly to arguments,
+  with a size-cutoff at which they will instead by returned via a hidden sret argument.
```
will instead be returned

JeffBezanson commented:
Thanks for adding the devdocs. That is very helpful.

ararslan commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+Note that extern functions are handled separately,
+with names, via the usual symbol resolution mechanism in the linker.
+
+Note too that ccall functions are also handled separately,
```
Perhaps code format ccall, i.e. `ccall`?

ararslan commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+When it returns and returns an `cgval.typ` set to `typ`.
+It'll cast the object to the requested representation.
+It'll make boxes, allocate stack copies, and compute tagged unions as
+needed to perform the request.
```
These sentences could be combined, e.g. "It will cast the object to the requested representation, making boxes, allocating stack copies, and computing tagged unions as needed."

vtjnash replied:
I'm not sure that necessarily improves readability – it just seems like a even trade, imo.

tkelman replied:
if these later "it will" sentences are supposed to be connected to the "when it returns" fragment, that could be made clearer either in sentence structure or markdown formatting

ararslan commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+- emit_sizeof
+- boxed
+- unbox
+- specialized cc-ret
```
Would be nice if you could code format these

vtjnash replied:
I'm generally reluctant to code-format every word that also appears in code. In part because it is not really standard grammar, and in part because I try to avoid having every other word be bracketed :P

ararslan replied:
I just mean the things that only refer to functions or other code-specific thingamabobs that aren't part of normal prose. But if you're happy with it as it is, that's fine with me.

ararslan commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+thus avoiding the cost of re-allocating a box,
+while maintaining the ability to efficiently handle union-splitting based on the low bits.
+
+It is guaranteed that `byte & 0x7f` is an exact test for the type,
```
I think either "since" should follow the comma, or the comma should be replaced by a semicolon.

ararslan commented on doc/src/index.md:
```diff
@@ -72,6 +72,7 @@
   * Documentation of Julia's Internals
       * [Initialization of the Julia runtime](@ref)
       * [Eval of Julia code](@ref)
+      * [High-level Overview of the Native-Code Generation Process](@ref)
```
I might put this after the memory layout section, since that and the preceding sections are helpful to understand prior to reading this (or at least it seems that way to me)

StefanKarpinski commented:
I've changed my review to "approve" on the premise that these docs are incomplete and when this is all done, @vtjnash will spell out the first instance of each acronym and link to the wikipedia page.

ararslan commented on doc/src/devdocs/compiler.md:
```diff
@@ -0,0 +1,119 @@
+# High-level Overview of the Native-Code Generation Process
+
+
+    <placeholder>
```
Are you planning to include something here in this PR or in a subsequent one?

ararslan replied:
Subsequent one it is 😄

vtjnash replied:
Yeah, I put it there so that nobody could complain that the document was immensely incomplete at describing codegen, and just jumps to explain a few small areas. It almost worked :)

maleadt commented:
@vtjnash: what needs to happen for that `/*TODO: min_align*/1` -- where does the alignment information need to come from?

I'm debugging an issue where a hot `size` on a custom array has alignment 1, which is very costly on GPU:

```julia
@inline foo(t) = 1 < 2 ? t[1] : 1   # might seem idiotic, but resembles Base.size
bar(t) = foo(t)
code_llvm(bar, (NTuple{1,Int},))

%2 = load i64, i64* %1, align 1
```

(strange how this triggers different code paths, but that's in part due to #17880 I guess)

vtjnash commented:
Likely need to track where the pointer came from in the jl_cgval_t, or make some conservative estimates like we do everywhere else. I didn't do that already mostly because x86 doesn't care.
